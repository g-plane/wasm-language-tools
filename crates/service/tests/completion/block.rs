use super::*;
use insta::assert_json_snapshot;
use lspt::{Position, Uri};
use wat_service::LanguageService;

#[test]
fn blocks() {
    let uri = "untitled:test".to_string();
    let source = "
(module
  (func
    (block
      (block $a
        (block
          (block
            (block $b
              (br_table ))))))
    (block)
    (block $c)))
";
    let mut service = LanguageService::default();
    service.commit(uri.clone(), source.into());
    let response = service.completion(create_params(uri, Position { line: 8, character: 24 }));
    assert_json_snapshot!(response);
}

#[test]
fn blocks_following_int_idx() {
    let uri = "untitled:test".to_string();
    let source = "
(module
  (func
    (block
      (block $a
        (block
          (block
            (block $b
              (br_table 1))))))
    (block)
    (block $c)))
";
    let mut service = LanguageService::default();
    service.commit(uri.clone(), source.into());
    let response = service.completion(create_params(uri, Position { line: 8, character: 25 }));
    assert_json_snapshot!(response);
}

#[test]
fn blocks_following_dollar() {
    let uri = "untitled:test".to_string();
    let source = "
(module
  (func
    (block
      (block $a
        (block
          (block
            (block $b
              (br_table $))))))
    (block)
    (block $c)))
";
    let mut service = LanguageService::default();
    service.commit(uri.clone(), source.into());
    let response = service.completion(create_params(uri, Position { line: 8, character: 25 }));
    assert_json_snapshot!(response);
}

#[test]
fn blocks_following_ident() {
    let uri = "untitled:test".to_string();
    let source = "
(module
  (func
    (block
      (block $a
        (block
          (block
            (block $b
              (br_table $a))))))
    (block)
    (block $c)))
";
    let mut service = LanguageService::default();
    service.commit(uri.clone(), source.into());
    let response = service.completion(create_params(uri, Position { line: 8, character: 26 }));
    assert_json_snapshot!(response);
}

#[test]
fn multiple_indexes() {
    let uri = "untitled:test".to_string();
    let source = "
(module
  (func
    (block
      (block $a
        (block
          (block
            (block $b
              (br_table $b 1 ))))))
    (block)
    (block $c)))
";
    let mut service = LanguageService::default();
    service.commit(uri.clone(), source.into());
    let response = service.completion(create_params(uri, Position { line: 8, character: 29 }));
    assert_json_snapshot!(response);
}

#[test]
fn block_type() {
    let uri = "untitled:test".to_string();
    let source = "
(module
  (type $t (func (result i32)))
  (func
    (loop (type $t)
      (block $a
        (block (param i32 i32) (result i32)
          (block
            (if $b (result i32 i32)
              (br_table ))))))
    (block)
    (block $c)))
";
    let mut service = LanguageService::default();
    service.commit(uri.clone(), source.into());
    let response = service.completion(create_params(uri, Position { line: 9, character: 24 }));
    assert_json_snapshot!(response);
}
