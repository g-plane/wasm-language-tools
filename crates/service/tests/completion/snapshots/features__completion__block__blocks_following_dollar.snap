---
source: crates/service/tests/completion/block.rs
expression: response
---
[
  {
    "label": "$b",
    "labelDetails": {
      "description": "[]"
    },
    "kind": 6,
    "detail": "(block $b)",
    "textEdit": {
      "range": {
        "start": {
          "line": 8,
          "character": 24
        },
        "end": {
          "line": 8,
          "character": 25
        }
      },
      "newText": "$b"
    }
  },
  {
    "label": "1",
    "labelDetails": {
      "description": "[]"
    },
    "kind": 6,
    "detail": "(block)",
    "textEdit": {
      "range": {
        "start": {
          "line": 8,
          "character": 24
        },
        "end": {
          "line": 8,
          "character": 25
        }
      },
      "newText": "1"
    }
  },
  {
    "label": "2",
    "labelDetails": {
      "description": "[]"
    },
    "kind": 6,
    "detail": "(block)",
    "textEdit": {
      "range": {
        "start": {
          "line": 8,
          "character": 24
        },
        "end": {
          "line": 8,
          "character": 25
        }
      },
      "newText": "2"
    }
  },
  {
    "label": "$a",
    "labelDetails": {
      "description": "[]"
    },
    "kind": 6,
    "detail": "(block $a)",
    "textEdit": {
      "range": {
        "start": {
          "line": 8,
          "character": 24
        },
        "end": {
          "line": 8,
          "character": 25
        }
      },
      "newText": "$a"
    }
  },
  {
    "label": "4",
    "labelDetails": {
      "description": "[]"
    },
    "kind": 6,
    "detail": "(block)",
    "textEdit": {
      "range": {
        "start": {
          "line": 8,
          "character": 24
        },
        "end": {
          "line": 8,
          "character": 25
        }
      },
      "newText": "4"
    }
  },
  {
    "label": "5",
    "labelDetails": {
      "description": "[]"
    },
    "kind": 6,
    "detail": "(func)",
    "textEdit": {
      "range": {
        "start": {
          "line": 8,
          "character": 24
        },
        "end": {
          "line": 8,
          "character": 25
        }
      },
      "newText": "5"
    }
  }
]
